### PR TITLE
Avoid reading emails from other domains (info@foo vs info@bar)

### PR DESCRIPTION
**Pull request by blaxter**
There are two fixes here, one is mayor and the other is minor. Both only applies for multidomain sogo scenario and are issues related with cached data related with logged users.

* Avoid creating cache entries like `user+attributes` (in case you logged in as, e.g., `user@foo.com`) to avoid reading emails from another user with same name but from different domain (e.g. `user@bar.com`).
* Avoid creating cache entries like `user@domain@domain+attributes`, no need to create that, there was being dup info (same contents that `user@domain+attributes` entries, depending of the method one of those was being read/written).

I'll let as exercise to the reader to determine which one is the mayor and which the minor one.

Suggested change to `NEWS` under `bug fix` section:
`Fixed reading emails from another user with same uid but from a different domain`

How To test this:
* Create two users with same uid but different domains.
* Log in sogo with `user 1`, open another browser and log in as `user 2`.
* Refresh `user 1` mails and also `user 2`'s (surprise mothe... I mean oh! mails from another user).

### CODE REVIEW

**Comment by extrafu** [comment]
Please update your patch, since we modified parts of the exact same code yesterday.

**Comment by blaxter** [comment]
cherrypicked extrafu's commit related with the second fix, this way there won't be conflicts in merges to inverse's repo

File: SoObjects/SOGo/SOGoUserManager.m [hunk 1]
```diff
@@ -372,10 +373,22 @@ - (NSString *) getExternalLoginForUID: (NSString *) uid
...
-      login = [dd forceExternalLoginWithEmail] ? [self getEmailForUID: uid] : uid;
+
+      if ([dd forceExternalLoginWithEmail])
+        {
+          sd = [SOGoSystemDefaults sharedSystemDefaults];
+          if ([sd enableDomainBasedUID])
+            {
+              // On multidomain environment we must use uid@domain
+              // for getEmailForUID method
+              login = [NSString stringWithFormat: @"%@@%@", uid, domain];
```
Comment (extrafu) [inline SoObjects/SOGo/SOGoUserManager.m]: I think we should check if UID here doesn't already have a @ - like all other checks we have.

Comment (blaxter) [inline SoObjects/SOGo/SOGoUserManager.m]: It's impossible at this point that `uid` has the domain (and if it does, there is another bug or design problem elsewhere). This method is called using `loginInDomain` as input parameter which should return the username part.

Personally I prefer to not include a check that could mask another issue.

**Comment by extrafu** [comment]
Can you do a PR on SOGo so I can merge it?

**Comment by blaxter** [comment]
@extrafu sure, I'll rebase on top of inverse/master first thing in the morning and send you the pr

**Comment by extrafu** [comment]
Any chance of doing it in the next hour? I could then include it in the upcoming nightly builds. Shouldn't take more than 5 mins :)

**Comment by sixstone-qq** [comment]
@blaxter is it meaningful to cherry-pick this one as well? https://github.com/inverse-inc/sogo/commit/4f303f8aa26046ecae916c457839d84622b260bb as @extrafu suggested.

File: SoObjects/SOGo/SOGoUserManager.m [hunk 2]
```diff
@@ -475,27 +488,22 @@ - (BOOL) checkLogin: (NSString *) _login
...
-  failedCount = [[SOGoCache sharedCache] failedCountForLogin: username];
-  dd = [SOGoSystemDefaults sharedSystemDefaults];
-
-  //
   // We check the fail count per user in memcache (per server). If the
   // fail count reaches X in Y minutes, we deny immediately the
   // authentications for Z minutes
-  //
+  failedCount = [[SOGoCache sharedCache] failedCountForLogin: username];
+  dd = [SOGoSystemDefaults sharedSystemDefaults];
```
Comment (sixstone-qq) [inline SoObjects/SOGo/SOGoUserManager.m]: Why don't you move `dd` assignment to the `if` block? it is used only inside there...

**Comment by sixstone-qq** [comment]
Apart from these two minor comments, it works like a charm :+1: :confetti_ball:

File: SoObjects/SOGo/SOGoUserManager.m (hunk 2 above)
Comment (blaxter) [inline SoObjects/SOGo/SOGoUserManager.m]: this will change once we merge inverse/master so not really important, though I agree with your suggestion

**Pull request merged**